### DISCUSSION TITLE
NEXTAUTH_URL Warning

### DISCUSSION THREAD

**Discussion by masnwilliams** [nextauthjs/next-auth#4220] [Help]
Hello! I recently moved my environment variables to be controlled by Vercel and then used `vercel pull` to get the latest version of them.

This is now resulting in this warning:
```
[next-auth][warn][NEXTAUTH_URL] 
https://next-auth.js.org/warnings#nextauth_url
```

I have `NEXTAUTH_URL` and `VERCEL_URL` filled in with `http://localhost:3000`.

I am not sure what the problem is. Any help would be greatly appreciated.

**Comment by ChrisB007** [comment]
Having same issue

**Comment by momotofu** [comment]
Same here. I recently migrated from Next-Auth V3 to V4. Everything seems to be functioning normally, except for this error. I've also verified that the environment variables `NEXTAUTH_URL` and `VERCEL_URL` are assigned `http://localhost:3000`.

**Comment by momotofu** [comment]
I found that Next-Auth is checking for the `host` property and the `req` object in `[...nextauth].ts`. The reason this error is thrown is because `host` is `undefined`. 

See the source code here: https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/core/lib/assert.ts#L47

The `host` property isn't declared on `import type { NextApiRequest, ... } from 'next'` so maybe this is a bug? Or is an warning unrelated to NextJS?

**Comment by markflorkowski** [comment]
Seems like removing all `VERCEL_` env vars from the .env you pull from vercel makes the warning go away. ~~Haven't tracked down which one in particular is causing the issue though.~~

Seems to be specifically having `VERCEL="1"` causes the error to come up

**Reply by markflorkowski** [reply]
As a workaround -- If you manually define a `VERCEL` environment variable for development in your Vercel project with an empty value, the warning does not appear

<img width="591" alt="image" src="https://user-images.githubusercontent.com/11494384/161697154-014d1815-4407-4049-968b-f1caec72ac47.png">

**Reply by markflorkowski** [reply]
Pretty sure this is the offending line of code: https://github.com/nextauthjs/next-auth/blob/0d54170e831fdd873d1aeb59c69ba33d02dbb4eb/packages/next-auth/src/jwt/index.ts#L73

**Reply by JoeRall** [reply]
@markflorkowski  - Removing `Vercel="1"` is a nice way to get rid of the warning. 

It looks like the detect-host function is ultimately responsible for this issue.

https://github.com/nextauthjs/next-auth/blob/2469e44572f23f709fa8c5c65c6b7a4eb2383e9f/packages/next-auth/src/utils/detect-host.ts#L4

Detect host function checks for that env variable, assumes the code is running on Vercel and returns the forwarded host. When running locally, this is empty.  @momotofu, thanks for showing where the assertion happens.

I wonder if that function should check for `NEXTAUTH_URL` and fallback to the `Vercel` forwarded host. This would allow next-auth specific configuration to take precedence over platform specific hooks instead of the other way around. Just an idea.

**Reply by ChristianIvicevic** [reply]
I've noticed that behavior as well. The use-case I have is that disabled the Git integration of Vercel with GitHub in favor of the Vercel CLI in the GitHub workflows. Within the script I do a `vercel pull --environment=production` prior to building. Since the respective env file contains a bunch of `VERCEL_` in addition to a `VERCEL` variable the NextAuth package is really having a hard time.

Usually `NEXTAUTH_URL` is set to `VERCEL_URL` should that be defined. However if `NEXTAUTH_URL` isn't explicitly set and `VERCEL_URL` is set to the empty string (after pulling the variables) Next.js itself fails the build process with an Invalid URL error during the final step where the built output is checked.

My current workaround is to pull the variables and apply the following script in the pipeline to remove all Vercel variables:
```javascript
const envfile = require('envfile');
const fs = require('fs');
const path = require('path');

(() => {
  const envPaths = [
    path.join(__dirname, '..', '.vercel', '.env.production.local'),
    path.join(__dirname, '..', '.vercel', '.env.preview.local'),
    path.join(__dirname, '..', '.vercel', '.env.development.local'),
    path.join(__dirname, '..', '.env'),
  ];
  envPaths
    .filter((filePath) => fs.existsSync(filePath))
    .forEach((filePath) => {
      const file = fs.readFileSync(filePath);
      const config = envfile.parse(file.toString());
      const fixedConfig = Object.entries(config)
        .filter(([key]) => !key.startsWith('VERCEL'))
        .reduce((obj, [key, value]) => ({ ...obj, [key]: value }), {});
      fs.writeFileSync(filePath, envfile.stringify(fixedConfig));
    });
})();
```
Basically the URL handling within the library is really difficult to understand as it's not easy in the first place to handle all environments correctly.

**Comment by officialrajdeepsingh** [comment]
NEXTAUTH_URL is the canonical URL of your site.

for development, NEXTAUTH_URL looks like
========================================

`NEXTAUTH_URL=http://localhost:3000/`

for production, NEXTAUTH_URL looks like
=======================================

`NEXTAUTH_URL=http://officialrajdeepsingh.dev`

Read more about NEXTAUTH_URL details of docs 
https://next-auth.js.org/configuration/options

**Reply by Nezo96** [reply]
Worked for me.

**Reply by Me-JJ** [reply]
<img width="968" alt="Screenshot 2023-08-27 at 4 05 03 PM" src="https://github.com/nextauthjs/next-auth/assets/98110216/796861ef-e5f5-4c4b-91a2-dc2e474fcd8e">
**I correctly added NEXTAUTH_URL, but the warning still appears.**

**Comment by Raff1010X** [comment]
Please set it in your .env file. - https://next-auth.js.org/warnings

Maybe simply add ".env" file in project folder with: "NEXTAUTH_URL=http://localhost:3000" and this will solve [next-auth][warn][NEXTAUTH_URL] warning.

Try it and give a reply.

**Reply by binodbastola007** [reply]
NOT SOLVED

**Comment by smzm** [comment]
### What worked for me :

After becoming sure about setting `NEXTAUTH_URL=http://localhost:3000` inside `.env` file, I change the `fetch` inside `authorize`  function like below : 
```js
async authorize(credentials) {
        const res = await fetch(process.env.NEXTAUTH_URL + '/api/login', {           
             // ...
          }),
        });
```

**Comment by kartik-bhandari** [comment]
write this in your .env file if you are in development
`NEXTAUTH_URL="http://localhost:3000/"`

change the link if you are in production

**Reply by leitel98** [reply]
how can you change that in production?

**Reply by deep231w** [reply]
dude i fixed my code by shift the .env from apps to root folder of nextjs

**Comment by ormartins01** [comment]
Aqui resolveu!

**Reply by achahboune** [reply]
how i m using docker deplyment option

**Comment by fritz-sport-ecommerce** [comment]
use node v21.6.2 solve my problem

**Comment by deep231w** [comment]
![image](https://github.com/user-attachments/assets/8e4eff74-d7a8-407e-8988-9120407616f8)
Just move youre .env file to nextapps root directory

**Comment by extinctsion** [comment]
I am also facing same issue when I provide `NEXTAUTH_URL="http://abc.netlify.app"`, it does not let me signin on netlify or `NEXTAUTH_URL="http://abc.vercel.app"` vercel. But when I set env variable as `NEXTAUTH_URL="http://localhost:3000"
` it works but on signout, it redirects me to `http://localhost:3000/`. I want to redirect it to `http://abc.netlify.app`
